src/autosave.js

```diff
@@ -0,0 +1,288 @@
+		// to warn before full page reload and this event cannot be dispatched manually.
+		/* istanbul ignore next */
+		this._domEmitter.listenTo( window, 'beforeunload', ( evtInfo, domEvt ) => {
+			if ( pendingActions.isPending ) {
+				domEvt.returnValue = pendingActions.first.message;
+			}
+		} );
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	destroy() {
+		// There's no need for canceling or flushing the throttled save, as
+		// it's done on the editor's destroy event with the highest priority.
+
+		this._domEmitter.stopListening();
+		super.destroy();
+	}
+
+	/**
+	 * Invokes remaining `_save` method call.
+	 *
+	 * @protected
+	 */
+	_flush() {
+		this._throttledSave.flush();
+	}
+
+	/**
+	 * If the provider is set and new document version exists,
+	 * `_save()` method creates a pending action and calls `provider.save()` method.
+	 * It waits for the result and then removes the created pending action.
+	 *
+	 * @private
+	 */
+	_save() {
+		const version = this.editor.model.document.version;
+
+		// Marker's change may not produce an operation, so the document's version
+		// can be the same after that change.
+		if ( !this.provider || version < this._lastDocumentVersion ) {
+			this._decrementCounter();
+
+			return;
+		}
+
+		this._lastDocumentVersion = version;
+
+		Promise.resolve( this.provider.save() )
+			.then( () => {
+				this._decrementCounter();
+			} );
+	}
+
+	/**
+	 * Increments counter and adds pending action if it not exists.
+	 *
+	 * @private
+	 */
+	_incrementCounter() {
+		this._saveActionCounter++;
+
+		if ( !this._action ) {
+			const pendingActions = this.editor.plugins.get( PendingActions );
+			this._action = pendingActions.add( 'Saving in progress.' );
+		}
+	}
+
+	/**
+	 * Decrements counter and removes pending action if counter is empty,
+	 * which means, that no new save action occurred.
+	 *
+	 * @private
+	 */
+	_decrementCounter() {
+		this._saveActionCounter--;
+
+		if ( this._saveActionCounter === 0 ) {
+			const pendingActions = this.editor.plugins.get( PendingActions );
+			pendingActions.remove( this._action );
+			this._action = null;
+		}
+	}
+}
+
+/**
+ * @interface module:autosave/autosave~SaveProvider
+ */
+
+/**
+ * Method that will be called when the data model changes. It might return a promise (e.g. in case of saving content to the database),
+ * so the `Autosave` plugin will wait for that action before removing it from the pending actions.
+ *
+ * @method #save
+ * @returns {Promise.<*>|undefined}
+ */
+
+/**
+ * Throttle function - a helper that provides ability to specify minimum time gap between calling the original function.
+ * Comparing to the lodash implementation, this provides an information if calling the throttled function will result in
+ * calling the original function.
+ *
+ * @private
+ * @param {Function} fn Original function that will be called.
+ * @param {Number} wait Minimum amount of time between original function calls.
+ */
+function throttle( fn, wait ) {
```

Comment:
I would move throttle function to the separate filer and introduce tests for it.

Reply:
I wanted to make it a private util. Maybe I could export the function from that file?

Reply:
But what is the point in keeping it private?